src/smirnoffio.py

```diff
@@ -138,6 +138,44 @@ def build_pid(self, element, parameter):
             logger.info("Minor warning: Parameter ID %s doesn't contain any SMIRKS patterns, redundancies are possible\n" % ("/".join([InteractionType, parameter])))
             return "/".join([ParentType, InteractionType, parameter])
 
+def assign_openff_parameter(ff, new_value, pid):
+    """
+    Assign a SMIRNOFF parameter given the openforcefield.ForceField object, the desired parameter value,
+    and the parameter's unique ID.
+    """
+    # Split the parameter's unique ID into four fields using a slash:
+    # Input: ProperTorsions/Proper/k1/[*:1]~[#6X3:2]:[#6X3:3]~[*:4]
+    # Output: ProperTorsions, Proper, k1, [*:1]~[#6X3:2]:[#6X3:3]~[*:4]
+    # The first, third and fourth fields will be used for parameter assignment.
+    # We use "value_name" to describe names of individual numerical values within a single parameter type
+    # e.g. k1 in the above example.
+    (handler_name, tag_name, value_name, smirks) = pid.split('/')
+
+    # Get the OpenFF parameter object
+    parameter = ff.get_parameter_handler(handler_name).parameters[smirks]
+
+    if hasattr(parameter, value_name):
+        # If the value name is an attribute of the parameter then we set it directly.
+        unit = getattr(parameter, value_name).unit
+        setattr(parameter, value_name, new_value*unit)
```

Comment:
I think this is the "API" call for changing the parameter values, consistent with the [example](https://github.com/openforcefield/openforcefield/blob/master/examples/forcefield_modification/ManipulateParameters.ipynb)
However this is not exactly "clean", as `value_name` can be other things that mistakenly override an unexpected attribute of this object. Also, I think the "k1", "k2" are intrinsic parameter names, so `parameter.k = [k1, k2]` may not be "safe" (i.e. if in the future there is a k0 before k1, or there is k1,k3 but skipped k2).
All of the above potential issues can be solve by an improved API in OpenFF toolkit, for example:
```
> parameter.get_name_values()
{
    'k1': 1.0,
    'periodicity1': 1.0,
    'k2': 3.0,
}

> parameter.get_value('k1')
1.0

> parameter.set_value('k1', 2.0)
> parameter.get_value('k1')
2.0

> parameter.set_value('k_random', 2.0)
ValueError: value with name "k_random" does not exist.
```
@j-wags

Reply:
> However this is not exactly "clean", as value_name can be other things that mistakenly override an unexpected attribute of this object.

Could you elaborate? We're trying to enumerate the scope of allowed parameter names so we don't run into trouble with XML/Python/other technology variable name restrictions. This could be useful input for the discussion here -- "Which attribute names should we allow?" https://github.com/openforcefield/openforcefield/issues/338 

> Also, I think the "k1", "k2" are intrinsic parameter names, so `parameter.k = [k1, k2]` may not be "safe" (i.e. if in the future there is a k0 before k1, or there is k1,k3 but skipped k2).

In the SMIRNOFF spec, we've been thinking about how to avoid any sort of misinterpretation along these lines. Right now, [indexed attributes must start at 1, and be consecutive](https://open-forcefield-toolkit.readthedocs.io/en/latest/smirnoff.html#propertorsions). There must also be an equal number of `k#`, `periodicity#` and `phase#` values in torsions (`idivf#` is a different beast, but that's just going to be tricky no matter what we do). 

> `> parameter.get_name_values()`

That's a good suggestion. Does the [`parameter.to_dict()` function](https://open-forcefield-toolkit.readthedocs.io/en/latest/api/generated/openforcefield.typing.engines.smirnoff.parameters.ParameterType.html#openforcefield.typing.engines.smirnoff.parameters.ParameterType.to_dict) come close to satisfying this need? If not, we can consider other options. 

> `> parameter.get_value('k1')`

Hopefully we can access for that in, using the `getattr`/`setattr` functions. I'll try to review [Andrea's PR](https://github.com/openforcefield/openforcefield/pull/377) on that ASAP

Reply:
The question of how to _add new torsion terms_ still remains. I've opened a spot to discuss this: https://github.com/openforcefield/openforcefield/issues/378

In the current implementation, I think adding new terms would require extending the list of `k` values (along with `periodicity` and `phase`). But we probably want to have a friendlier option, since the interdependence between the separate lists is quite important.

Reply:
Sure I'm glad to hear that these improvements are on the schedule.